install/0000_90_machine-api-operator_04_alertrules.yaml

```diff
@@ -22,16 +22,26 @@ spec:
             severity: warning
           annotations:
             message: "machine {{ $labels.name }} does not have valid node reference"
-    - name: machine-with-no-running-phase
+    - name: machine-not-yet-deleted
       rules:
-        - alert: MachineWithNoRunningPhase
+        - alert: MachineNotYetDeleted
           expr: |
-            (mapi_machine_created_timestamp_seconds{phase!="Running"}) > 0
+            (mapi_machine_created_timestamp_seconds{phase=="Deleting"}) > 0
           for: 60m
           labels:
             severity: warning
           annotations:
-            message: "machine {{ $labels.name }} is in phase: {{ $labels.phase }}"
+            message: "machine {{ $labels.name }} is having trouble deleting"
+    - name: machine-not-yet-running
+      rules:
+        - alert: MachineNotYetRunning
```

Comment:
This alert is inaccurate.  There is also a failed phase, and that could have happened after it became running in some circumstances.

Reply:
If that happens, won't MHC quickly (certainly within 60m) set it to Deleting and try to blow it away?

Though philosophically maybe we're trying not to tie these to MHC behavior.

Perhaps a third alert that just deals with Failed (and exclude that state from these two)?

Reply:
IMO we shouldn't tie this to MHC as some users may not be using it.

I'd be happy with excluding failed from MachineNotYetRunning and having a separate alert that says hey you have a failed machine and you haven't done anything about it, you should probably fix that, kinda thing

@elmiko Interested to know your thoughts on this bit in particular?

Reply:
> IMO we shouldn't tie this to MHC as some users may not be using it.

+1

> I'd be happy with excluding failed from MachineNotYetRunning and having a separate alert that says hey you have a failed machine and you haven't done anything about it, you should probably fix that, kinda thing

i tend to agree with this notion. i like the idea of the alerts being granular enough to specify what happened and what the user should do next. i'm trying to think about this from the point of someone who has just received either alert (machine failed for too long, or machine not running after a long time), and what they would do next. is there a notable difference in triaging and fixing between these two alerts? i feel like in either case the user is going to want to somehow remove the instance and replace it, but i'm curious if there is some aspect i am missing.

Reply:
> is there a notable difference in triaging and fixing between these two alerts?

I'm coming at this from SRE-P's point of view. In our case MHC is always present; so if these alerts fire with their current configuration of 60m, it means MHC didn't/couldn't fix things. I don't know what scenarios that could entail, but at a guess, my action should be to force-delete the node and/or machine? BUT if the alert was MachineNotYetDeleted, I would want to force-drain the node first. I imagine that last thing applies with or without MHC.

Reply:
> I don't know what scenarios that could entail, but at a guess, my action should be to force-delete the node and/or machine? BUT if the alert was MachineNotYetDeleted, I would want to force-drain the node first. I imagine that last thing applies with or without MHC.

that makes sense to me. i should have been more specific about which two alerts i was talking about, i was trying to draw some distinction between the suggested alerts that @JoelSpeed was talking about (eg one for failed machines and one for non-failed but also non-running machines).

Reply:
Typically, one for failed would be highlighting issues where you can't boot VMs due to some unknown configuration error, or the VM for some reason went away for some unknown reason.

This to me is different to the deleting issue where we can't drain and warrants its own alert.

Whether an MHC is present or not, you can still get machines into Failed states and we should cater for these cases as our alerts are also used on clusters without MHC.

Reply:
Failing to drain is likely to be an in-cluster issue, e.g. user workloads are guarded by aggressive PDBs.  Failing to create is very unlikely to be an in-cluster, user-workload issue.  It might be a provider issue (e.g. "out of quota"), or it might be an in-cluster, admin issue (e.g. "borked Machine config").  Those are very different resolution paths.  And yeah, if you happen to be running MachineHealthChecks, either way you'll get a delete/replace.  That would blow through things like the PDB blockers, but would not help you grow your platform instance quota or fix your borked Machine config.  Would be really nice if there was an alert specific to quota exhaustion, but one step at a time ;).